### PR TITLE
Speed up Hessian matrix computation

### LINKED ISSUE #2182: Hessian filter implementation is very slow
Computing the largest eigenvalue of the Hessian matrix (`hessian_matrix` in `skimage/feature/corner.py`) on a 1250 x 1250 image takes about 3s on my machine, but is virtually instantaneous in Fiji, which isn't doing anything smart... Not sure whether `ndi.convolve` is to blame, but we can probably try to separate the kernels to use convolve1d, or other tricks.

### PR DESCRIPTION
**Pull request by jni**
Fixes #2182.

The Hessian filter can be separated into a (separable) Gaussian filtering followed by a gradient estimation. This method doesn't give the exact same numerical result, but the output correlation is ~99.5% with the existing implementation, 99.75% if you ignore the borders. This should be sufficient for all practical downstream uses.

### CODE REVIEW

**Comment by jni** [comment]
I should mention that the speedup is 17x for sigma=3.

**Comment by stefanv** [comment]
Beautiful!  N-D and a big speed-up.  @ahojnnes, I think you wrote the original code, if you would like to have a look?  I'm happy with the dependency bump to np 1.11.

**Comment by ahojnnes** [comment]
LGTM, but why is the numpy version bump necessary?

**Comment by sciunto** [comment]
@jni there are unexpected warnings with the min requirements

File: skimage/feature/corner.py [hunk 1]
```diff
@@ -136,43 +138,31 @@ def hessian_matrix(image, sigma=1, mode='constant', cval=0):
...
-    Hxx = ndi.convolve(image, kernel_xx, mode=mode, cval=cval)
-    Hxy = ndi.convolve(image, kernel_xy, mode=mode, cval=cval)
-    Hyy = ndi.convolve(image, kernel_yy, mode=mode, cval=cval)
+    gradients = np.gradient(gaussian_filtered)
+    axes = range(image.ndim)
+    H_elems = [np.gradient(gradients[ax0], axis=ax1)
+               for ax0, ax1 in combinations_with_replacement(axes, 2)]
 
-    return Hxx, Hxy, Hyy
+    if image.ndim == 2:
```
Comment (soupault) [inline skimage/feature/corner.py]: Any plans to add deprecation path to this?

Comment (ahojnnes) [inline skimage/feature/corner.py]: +1

Comment (jni) [inline skimage/feature/corner.py]: There's *so much code* that depends on this... I would like to leave deprecation for a later PR. I'll raise an issue when this gets merged.

**Comment by jni** [comment]
@ahojnnes re NumPy 1.11, the axis kwarg to np.gradient was only added in 1.11. Without it, we end up computing 2x as many gradients as we need.

**Comment by jni** [comment]
@sciunto yes I'm trying to track down where `np.rank` is called. Doesn't seem to be our code, but rather PyAMG... Hold, please... =)

File: skimage/feature/corner.py (hunk 1 above)
Comment (ahojnnes) [inline skimage/feature/corner.py]: @jni Fine with me.

**Comment by ahojnnes** [comment]
@jni Thanks for the clarification.

**Comment by sciunto** [comment]
Minor stuffs: 

* In the docstring of `structure_tensor`, missing `,  optional` for sigma.
* For `hessian_matrix_det`, I found a bit inconsistent that `sigma` is not an optional parameter.

**Comment by sciunto** [comment]
Would you like to fix this as well, please?

**Comment by ahojnnes** [comment]
LGTM, appveyor error seems to be unrelated.

**Comment by JDWarner** [comment]
LGTM as well. Merging despite AppVeyor, which clearly doesn't like SciPy right now.

**Comment by stefanv** [comment]
Good work, thanks Juan.

**Pull request merged**